### DISCUSSION TITLE
Operations priority regression in Tasmota 14.3.0 and ESP32 (not on ESP8266)

### DISCUSSION THREAD

**Discussion by aldweb** [arendst/Tasmota#22636] [Q&A] [answered]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._

I have a very simple rule instruction which looks like:
ON VAR1#State DO IF (%uptime%>1) MEM1=(59*%MEM1%+%value%)/60 ENDIF ENDON
This one has worked forever on my various ESP8266 and ESP32 devices.
But, as from Tasmota 14.3.0, and only on ESP32 devices, the result stored in MEM1 is erroneous.
Example: MEM1 10 and VAR1 10 reports MEM1 98.333 (when it should report 10)

If I change the rule instruction to:
ON VAR1#State DO IF (%uptime%>1) MEM1=((59*%MEM1%)+%value%)/60 ENDIF ENDON
then the result stored in MEM1 is good.
Example: MEM1 10 and VAR1 10 reports MEM1 10 which is the expected value

This means that operations priority is not working fine in the latest Tasmota version.

In in user_config_override.h, I have set:
#define USE_EXPRESSION
#define SUPPORT_IF_STATEMENT

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [X] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [X] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [X] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [X] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [X] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [X] Device used (e.g., Sonoff Basic): ESP32 Thing (sparkfun)
- [x] Tasmota binary firmware version number used: 14.3.0
  - [ ] Pre-compiled
  - [X] Self-compiled
- [ ] Flashing tools used: _____
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [X] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:

20:49:43.847 CMD: Backlog Rule1; Rule2; Rule3
20:49:43.862 RSL: RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":72,"Free":439,"Rules":"ON VAR1#State DO IF (%uptime%>1) MEM1=(59*%MEM1%+%value%)/60 ENDIF ENDON"}}
20:49:44.095 RSL: RESULT = {"Rule2":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}
20:49:44.295 RSL: RESULT = {"Rule3":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}

```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [X] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:

21:05:48.135 CMD: mem1 10
21:05:48.136 SRC: WebConsole from 192.168.XXX.XXX
21:05:48.138 CMD: Grp 0, Cmd 'MEM', Idx 1, Len 2, Pld 10, Data '10'
21:05:48.139 CFG: CR 402/699, Busy 0
21:05:48.141 RSL: RESULT = {"Mem1":"10"}
21:05:48.339 CFG: Saved, Count 2961, Bytes 4096
21:05:51.645 CMD: var1 10
21:05:51.647 SRC: WebConsole from 192.168.XXX.XXX
21:05:51.648 CMD: Grp 0, Cmd 'VAR', Idx 1, Len 2, Pld 10, Data '10'
21:05:51.651 RSL: RESULT = {"Var1":"10"}
21:05:51.676 RUL: VAR1#STATE performs 'IF (17>1) MEM1=(59*10+10)/60 ENDIF'
21:05:51.677 SRC: Rule
21:05:51.679 CMD: Grp 0, Cmd 'IF', Idx 1, Len 31, Pld -99, Data '(17>1) MEM1=(59*10+10)/60 ENDIF'
21:05:51.682 RSL: RESULT = {"If":"Done"}
21:05:51.892 SRC: Backlog
21:05:51.894 CMD: Grp 0, Cmd 'MEM', Idx 1, Len 14, Pld -99, Data '=(59*10+10)/60'
21:05:51.896 CFG: CR 406/699, Busy 0
21:05:51.897 RSL: RESULT = {"Mem1":"98.333"}
21:05:52.322 CFG: Saved, Count 2962, Bytes 4096

```

### TO REPRODUCE
_Steps to reproduce the behavior:_

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by sfromis** [comment]
Looks like an issue is referencing var1 for the calculation, with very varied numbers being used. On an ESP32-C3, I'm getting output like:
```
23:10:23.566 RUL: VAR1#STATE performs 'IF (552>1) MEM1=(59*0.98+10)/60 ENDIF'
23:10:43.785 RUL: VAR1#STATE performs 'IF (553>1) MEM1=(59*9.64+10)/60 ENDIF'
23:10:56.803 RUL: VAR1#STATE performs 'IF (553>1) MEM1=(59*94.79+10)/60 ENDIF'
```
Same happens if I try to reference `%var1%` instead of `%value%`.

Hence, it does not look like a question of priority, but instead some issue with getting the value at all.

Note that since some time now, you do not need to build with explicit `USE_EXPRESSION` and `SUPPORT_IF_STATEMENT`, as these have become default for all builds.

**Comment by arendst** [comment]
Nice find.

On ESP32:
With ``mem1 10`` and ``var1 10`` and your rule ``ON VAR1#State DO IF (%uptime%>1) MEM1=(59*%MEM1%+%value%)/60 ENDIF ENDON`` which is handled like ``IF (8470>1) MEM1=(59*10+10)/60 ENDIF`` results in mem1 being 98.333 as a result from ``IF (8470>1) MEM1=(59*10*10)/60 ENDIF``.

Apparently the second 10 is not added but multiplied!

On ESP8266 it performs as expected and results in mem1 being 10.000

So it's not so much an operation priority issue but for some reason an addition is changed into a multuplication.

Let me see where this fails...

**Comment by arendst** [comment]
Got it.

Code ``operators[i++] = operators[i];`` works as expected ``operators[i] = operators[i +1];`` on esp8266. It doesn't on esp32 ;-(

Will fix ASAP.

**Comment by arendst** [comment, marked as answer]
Fixed in latest dev.

https://github.com/arendst/Tasmota/commit/620fade73c902352574aa9675c7a438764c18f19

**Reply by aldweb** [reply]
Good news, that was fast fixing!
